docs/proposals/repo-server-multi-tenancy.md

```diff
@@ -0,0 +1,200 @@
+
+The initial design assumption for the repository server was, that it will not have access to any confidential data (e.g. treat manifests and contents in Git as non-confidential), and as noted above, it does not even have access to the Kubernetes cluster it is running on. With this assumption, it wouldn't matter if a tool executed on behalf of user A could also read manifest data from user B, due to the semi-public nature of the manifests. However, this premise has changed over the course of time, and the Argo CD community came up with solutions for secrets management and has integrated tools like git-crypt, helm-secrets and others, all of which require a private key to be known inside the repository server. Furthermore, tools like Helm and Kustomize have evolved to allow all kinds of things, such as executing commands as plugins, refering to files outside the current repository, and more. This would make it rather easy to read any secret stored within the repo server's file system for users who can control the source of any application rendered by these tools.
+
+There have been some changes made recently to increase the difficulty for any tool executed on behalf of user A to gain access to resources of user B, including creating random, unpredictable file system path names. However, this is more _security by obscurity_ than anything else. Also, the Config Management Plugin (CMP) v2 approach implements some level of isolation, but it does not go far enough to provide a strict isolation required by tenants that do not trust each other.
+
+### Goals
+
+* Provide proper isolation for the repository server when Argo CD is setup in a scenario where different users/tenants do not trust each other.
+* Allow different parties (i.e. "tenants") the safe usage of tools that require secret data, such as a private key mounted into the repo server, to be available in order to render manifest data
+* Limit the impact of rogue or misbehaving tools
+* Open the door for more use cases in the future, like managing confidential data (such as private PGP keys) on a per-tenant (AppProject) basis
+
+### Non-Goals
+
+* There is no intention to replace CMP v2, but rather allow it to be leveraged in non-trusting multi-tenancy environments.
+* There are no intentions for architectural changes to the repository server itself, or the gRPC API for communication with 
```

Comment:
> or the gRPC API for communication with

for communication with what? I guess it should say 

```suggestion
* There are no intentions for architectural changes to the repository server itself, or the gRPC API for communication with the repository server
```

Reply:
Good catch! Merged.